### ISSUE TITLE
Proposal: Share service with namespaces

### ISSUE THREAD

**Issue by wienczny** [kubernetes/kubernetes#19122]
Namespaces are Kubernetes implementation of administration domains.
If you have a namespace for services that should be shared across namespaces you have to either resolve the the ip using the api or use the dns-addon.
I propose to add a sharing feature to services in namespaces:

- A boolean flag ```exported``` is added to the service definition.
or
- A list of names of namespaces is added to the service defintion.

The first solution should be easier to implement. The second one is more granular.

**Comment by thockin** [comment]
Can you explain what you think this flag would do in concrete terms?  Even
within a given namespace you use IP addresses or DNS to resolve Services,
so what would this change?

**Comment by wienczny** [comment]
As DNS is (currently #11599) an addon and not stable in operation, you should not have to rely it.
Alternatively you can use a static cluster IPs for you services. 
The third possiblilty is using service accounts and the kube-api
 
While inside the same namespace service IPs and ports are injected into the environment of started containers. It would be great if you could have services from another namespace available to you container.
Because the environment gets quite large if you use many services I propose a granular approach.
Huge environments should be another issue though.

**Comment by thockin** [comment]
On Mon, Dec 28, 2015 at 1:24 PM, Stephan Wienczny
<notifications@github.com> wrote:
>
> As DNS is (currently #11599) an addon and not stable in operation, you should not have to rely it.

Pardon?  In what sense is it not stable?  the vast vast vast majority
of kubernetes users rely on DNS, and in the eventual future it will
not be an addon.

> Alternatively you can use a static cluster IPs for you services.

That's a disaster operationally.

> The third possiblilty is using service accounts and the kube-api

This assumes you control all applications, which is why we went with
DNS - we can not make that assumption.

> While inside the same namespace service IPs and ports are injected into the environment of started containers. It would be great if you could have services from another namespace available to you container.

We hope to kill off env vars.  Env vars are terrible because they can
not be updated - if you start a service after a pod, you can not see
that variable.

> Because the environment gets quite large if you use many services I propose a granular approach.
> Huge environments should be another issue though.

We have some thoughts around explicitly importing other Services into
your own namespace, but it was more about DNS and network policy than
env vars.

**Comment by wienczny** [comment]
I agree with you concerning cluster IPs.

There are some issues I've got with the current DNS implementation:

Sometimes DNS requests timeout which might cause a lots of restarts.

```
kubectl --namespace=kube-system get pod
NAME                                   READY     STATUS    RESTARTS   AGE
kube-dns-v9-ziyjr                      4/4       Running   20         8d
                                                           ^^
```

The current Skydns implementation is hard to debug using the logs. I had to dig deeper to find the underlying problem there.

I've got some containers that don't contain a full linux environment. They are just a single binary with some data. For those containers getting DNS resolution to work without using anything from environment is quite hard. You had to compile the DNS IP into the application.

If you want a dynamic service resolution do you plan to have explicit service dependencies? So that you can have pods restartet if their dependencies change? If you have to prepare configuration files for a application this would be a requirement.

**Comment by thockin** [comment]
On Mon, Dec 28, 2015 at 4:57 PM, Stephan Wienczny
<notifications@github.com> wrote:
>
> I agree with you concerning cluster IPs.
>
> There are some issues I've got with the current DNS implementation:
>
> Sometimes DNS requests timeout which might cause a lots of restarts.
>
> kubectl --namespace=kube-system get pod
> NAME                                   READY     STATUS    RESTARTS   AGE
> kube-dns-v9-ziyjr                      4/4       Running   20         8d
>                                                            ^^

That's not normal - we should fix that, rather than tossing the baby
with the bathwater.

> The current Skydns implementation is hard to debug using the logs. I had to dig deeper to find the underlying problem there.

The current implementation will become more integrated over time -
it's a matter of priority and making time to work on it.

> I've got some containers that don't contain a full linux environment. They are just a single binary with some data. For those containers getting DNS resolution to work without using anything from environment is quite hard. You had to compile the DNS IP into the application.

Docker (and kubernetes) always configures /etc/resolv.conf to point to
the cluster DNS server.  Most/all language runtimes have some binding
to DNS to use resolv.conf to do DNS resolution.  We run many of our
own apps as "from scratch" containers with static binaries, but DNS
still works.

> If you want a dynamic service resolution do you plan to have explicit service dependencies? So that you can have pods restartet if their dependencies change? If you have to prepare configuration files for a application this would be a requirement.

Explicit service dependencies gives us an opportunity to optimize
and/or lock things down.  It's not a requirement for dynamic
resolution.  That works today (unless I misunderstand what you mean?).
Env vars are a last resort and are terrible for dynamic anything, so
lets ignore them.  If you write your app to do DNS lookups of the
services you depend on, you will get dynamic responses.  No restart
needed.

**Comment by wienczny** [comment]
>Explicit service dependencies gives us an opportunity to optimize
>and/or lock things down.  It's not a requirement for dynamic
>resolution.  That works today (unless I misunderstand what you mean?).
>Env vars are a last resort and are terrible for dynamic anything, so
>lets ignore them.  If you write your app to do DNS lookups of the
>services you depend on, you will get dynamic responses.  No restart
>needed.

DNS might be an improvement to env vars but It does not solve the signaling issue. Any application using the DNS for dynamic services has to repeatly poll the DNS for changes. This might work for your own applications but fails if you are using applications from the shelf. Is there a plan for a tool like confd bound to the service changes?

**Comment by thockin** [comment]
On Tue, Dec 29, 2015 at 3:51 PM, Stephan Wienczny
<notifications@github.com> wrote:
>
> Explicit service dependencies gives us an opportunity to optimize
> and/or lock things down. It's not a requirement for dynamic
> resolution. That works today (unless I misunderstand what you mean?).
> Env vars are a last resort and are terrible for dynamic anything, so
> lets ignore them. If you write your app to do DNS lookups of the
> services you depend on, you will get dynamic responses. No restart
> needed.
>
> DNS might be an improvement to env vars but It does not solve the signaling issue. Any application using the DNS for dynamic services has to repeatly poll the DNS for changes. This might work for your own applications but fails if you are using applications from the shelf. Is there a plan for a tool like confd bound to the service changes?

We have 2 modes of DNS for Services - "normal" and "headless".  Normal
services are the default - they give you a stable virtual IP that
doesn't need to be refreshed.  Headless services bypass the virtual
IP, but the assumption is that you opted in to that, so your app must
be under your control (e.g. refreshes properly).

I don't know anyone who has linked confd and kubernetes
Services/Endpoints API but it sounds pretty straightforward if you
REALLY need to go that way.

Are you actually experiencing a problem or just trying to grok the
system and asking good questions as you see them?

**Comment by wienczny** [comment]
I'm already using both service types. The headless ones are useful to resolve pod ips for my mysql-galera-nodes.
The problem I was starting this thread for was making the galera cluster available to customers using their own namespaces. As DNS is declared as "addon" I did not see it as best solution for the problem.
The initial problem was my understanding of term "addon" and issues with the dns as described above ;) Then I tried to understand how you designed it to be used.
As this is intended behavior I'll close this issue.
